Feature/115/computed var metadata api

asizemore opened the pull request:
Resolves #115 

## Details 

The current structure of "computedVariableMetadata" as defined in the data service looks like this:

```ComputedVariableMetadata:
    additionalProperties: false
    properties: 
      displayLabel:
        type: string[]
        required: false
      defaultRange:
        type: number[]
        maxItems: 2
        minItems: 2
        required: false
      vocabulary:
        type: string[]
        required: false
      isOrdinal: 
        type: boolean
        required: false
```

We're only using defaultRange and displayLabel for the near- and mid-future since planned computed vars so far are numeric, so only these properties are the two implemented and tested in this PR.

## Notes

I struggled with passing each property as an argument, or passing the whole list as an argument. I expect the computedVariableMetadata to gain more and more properties, which then combined with it being passed directly from the compute output swayed me towards just inputting the list itself. Still, not sure passing a list as a function argument is really proper so maybe it should be the other way...

asizemore commented:
Also is "defaultRange" the right term? Maybe should be "annotatedRange" to be more similar with var annotations?

d-callan commented:
ill look at this today. but i think displayRange is what the EDA* services are calling it.

i also think the list approach is the best for now. typically when you start to have those types of debates w yourself its a good sign your headed toward a new object. theres a few places in plot.data thats the case, and i think we can handle them all together at some point under some general 'refactor' issue. maybe when we switch to s4.

d-callan reviewed (commented):
pretty good. might have computed dates some day w ranges though? might need to update that here and in the api (would mean a string type rather than number). 

i also double checked the display range api for variable metadata in the subset service. its two separate properties, rather than one w two values. the two are 'displayRangeMin' and 'displayRangeMax'. i think we should match that. i also think were using 'displayName' rather than 'displayLabel'. so that i think needs updating here and in the data service api as well.

d-callan commented:
i guess the other thing is to make sure in the data service were actually passing this through from the compute output to plot.data. maybe open a second PR there?

asizemore commented:
Great. Updated language to use "displayName", "displayRangeMin", and "displayRangeMax".
Also updated output of range min and max to be strings. The compute var details will have if it's a date or number so we can use that in the front end to convert as necessary.

asizemore commented:
Heading to the data service to make a corresponding PR. Will link it here when done!

asizemore commented:
See also microbiomeComputations [PR 5](https://github.com/VEuPathDB/microbiomeComputations/pull/5) and data service [PR 134](https://github.com/VEuPathDB/EdaDataService/pull/134)

asizemore commented:
After discussion, we're planning to add collection var info to the computed var metadata. I'll add that to this PR. IP

asizemore commented:
New output example:

```
"collectionVariable": {
          "collectionType": "abundance",
          "collectionVariablePlotRef": "facetVariable2",
          "collectionValuePlotRef": "yAxisVariable",
          "collectionVariableDetails": [
            {
              "variableId": "contB",
              "entityId": "entity"
            },
            {
              "variableId": "contA",
              "entityId": "entity"
            }
          ]
        }
```

asizemore commented:
This PR also addresses #116 

With microbiomeComputations [PR 6](https://github.com/VEuPathDB/microbiomeComputations/pull/6), now writing collectionVariableDetails and other info in `computedVariableMetadata`. Apps will return the collectionType, then plot.data will add any collection var plot info. 

Currently keeping plot.data's collectionVar separate from computed var metadata as long as possible, since this likely a short-term solution.

See also data service [PR](https://github.com/VEuPathDB/EdaDataService/pull/134) for updates.

d-callan commented on R/class-plotdata-beeswarm.R:
```diff
@@ -27,10 +27,14 @@ newBeeswarmPD <- function(.dt = data.table::data.table(),
-                                               'listVarPlotRef' = NULL,
-                                               'listVarDisplayLabel' = NULL),
+                                               'collectionVariablePlotRef' = NULL,
+                                               'collectionVarDisplayLabel' = NULL),
```
my only question is about the display label here. maybe i just missed it, but im not sure i saw it used anywhere. plus it probably makes the most sense for the client to find display labels given variable specs..

asizemore replied:
Sounds good. collectionVarDisplayLabel removed!
